codex/purchasing/states/started.nim

```diff
@@ -1,9 +1,20 @@
 import ../statemachine
+import ./error
+import ./finished
 
 type PurchaseStarted* = ref object of PurchaseState
 
-method enter*(state: PurchaseStarted) =
+method enterAsync*(state: PurchaseStarted) {.async.} =
   without purchase =? (state.context as Purchase):
     raiseAssert "invalid state"
 
-  purchase.future.complete()
+  let clock = purchase.clock
+  let market = purchase.market
+  let request = purchase.request
+
+  try:
+    await clock.waitUntil(await market.getRequestEnd(request.id))
+  except CatchableError as error:
```

Comment:
Is it possible to narrow down this Error type a bit more? We may get a `JsonRpcProviderError` from ethers. Any others?

Reply:
Hard to say, I don't think that `nim-ethers` wraps all exceptions from its underlying libraries into `EthersError`s. Though, no matter what kind of error occurs here, we can handle them all the same: move to the `Error` state, so that the REST API can report the error to the user.